File: elixir/space-age/space_age.exs
defmodule SpaceAge do
  @type planet :: :mercury | :venus | :earth | :mars | :jupiter
                | :saturn | :neptune | :uranus
  @seconds_on_earth 31557600
  @seconds_by_planet %{mercury: @seconds_on_earth * 0.2408467,
                       venus: @seconds_on_earth * 0.61519726,
                       earth: @seconds_on_earth,
                       mars: @seconds_on_earth * 1.8808158,
                       jupiter: @seconds_on_earth * 11.862615,
                       saturn: @seconds_on_earth * 29.447498,
                       neptune: @seconds_on_earth * 164.79132,
                       uranus: @seconds_on_earth * 84.016846}

  @doc """
  Return the number of years a person that has lived for 'seconds' seconds is
  aged on 'planet'.
  """
  @spec age_on(planet, pos_integer) :: float
  def age_on(planet, seconds) do
    seconds / @seconds_by_planet[planet]
  end
end

File: haskell/space-age/src/SpaceAge.hs
module SpaceAge (Planet(..), ageOn) where

data Planet = Mercury
            | Venus
            | Earth
            | Mars
            | Jupiter
            | Saturn
            | Uranus
            | Neptune

earthSecondsPerYear :: Float
earthSecondsPerYear = 31557600

ageOn :: Planet -> Float -> Float
ageOn planet seconds = seconds / (earthSecondsPerYear * orbitalPeriod planet)

orbitalPeriod :: Planet -> Float
orbitalPeriod Mercury = 0.2408467
orbitalPeriod Venus   = 0.61519726
orbitalPeriod Earth   = 1.0
orbitalPeriod Mars    = 1.8808158
orbitalPeriod Jupiter = 11.862615
orbitalPeriod Saturn  = 29.447498
orbitalPeriod Uranus  = 84.016846
orbitalPeriod Neptune = 164.79132
